grpc-gcp/src/main/java/com/google/cloud/grpc/GcpManagedChannel.java

```diff
@@ -1189,6 +1459,36 @@ private ChannelRef tryCreateNewChannel() {
     return null;
   }
 
+  private void maybeDynamicUpscale() {
+    if (!isDynamicScalingEnabled || channelRefs.size() >= maxSize) {
+      return;
+    }
+
+    if ((totalActiveStreams.get() / channelRefs.size()) >= maxRpcPerChannel) {
+      dynamicUpscale();
+    }
+  }
+
+  private synchronized void dynamicUpscale() {
+    if (!isDynamicScalingEnabled || channelRefs.size() >= maxSize) {
+      return;
+    }
+
+    if ((totalActiveStreams.get() / channelRefs.size()) >= maxRpcPerChannel) {
+      createNewChannel();
+      scaleUpCount++;
```

Comment:
What if `(totalActiveStreams.get() / channelRefs.size()) >= maxRpcPerChannel` still happens after you create a new channel? In other words, what if you need to create multiple channels at the same time for scaling up?

Reply:
The condition you described is not quite possible because we perform this check for every upcoming RPC, hence we may only have a +1 RPC difference between the checks, we cannot jump to +100 RPC instantly.

But even if we jumped this would still be quite safe. The first check would create one channel and the very next RPC would create another one and so on until the condition is not satisfied. This part is synchronized, so we won't create more channels than we need.

Reply:
By synchronized do you mean the channel creation is in the critical path of a RPC, i.e., a RPC needs to wait for the channel creation? If so, will this be a performance issue?

Reply:
No, it shouldn't be a performance issue because we first check if we need to create a new channel and if we do, only then we acquire the lock, check again and create a channel if still needed. And channel creation is quite fast -- we don't wait until it become ready or anything like that.

Reply:
A channel creation is not fast, it can be hundreds of milliseconds. For example, in gRPC RR LB policy, gRPC will only assign RPCs to subchannels that are READY, and wait for subchannels that are connecting. In this way, channel creation will only cause latency at the beginning of the process, and as long as there is at least one connection, RPCs will not be blocked. Here it sounds not this case, and a RPC will need to wait for a channel creation even there are other channels that are READY, correct? If so, I am worried here, since this is a dynamical channel pool, channels can keep being created (and destroyed).

Reply:
We synced up office. The known risk is that RPCs could be blocked by the new channel creation (if fallback is enabled, only one RPCs will be blocked; if fallback is not enabled, multiple RPCs will be blocked because RPC assignment is based on the load of channel and the new channel that is being created has no load. Cloud Spanner can configure this behavior).

However, we think this risk is OK since:
1. The channel scale up and down should be rare (at most every a few minutes);
2. We have observability by correlating the channel scale up/down metric and the latency metric,